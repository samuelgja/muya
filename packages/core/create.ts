import { canUpdate, generateId } from './utils/common'
import type { Emitter } from './utils/create-emitter'
import { createEmitter } from './utils/create-emitter'
import { isEqualBase, isFunction, isSetValueFunction, isUndefined } from './utils/is'
// import { createScheduler } from './utils/scheduler'
import type { Cache, Callable, DefaultValue, IsEqual, Listener, SetValue } from './types'
import { context } from './subscriber'
import { createGlobalScheduler } from './utils/global-scheduler'

export const createScheduler = createGlobalScheduler()
interface RawState<T> {
  (): T
  id: number
  set: (value: SetValue<T>) => void
  emitter: Emitter<T>
  listen: Listener<T>
  destroy: () => void
  withName: (name: string) => RawState<T>
  stateName?: string
}

export type State<T> = {
  readonly [K in keyof RawState<T>]: RawState<T>[K]
} & Callable<T>

export function create<T>(initialValue: DefaultValue<T>, isEqual: IsEqual<T> = isEqualBase): State<T> {
  const cache: Cache<T> = {}

  function getValue(): T {
    if (isUndefined(cache.current)) {
      cache.current = isFunction(initialValue) ? initialValue() : initialValue
    }
    return cache.current
  }
  function resolveValue(value: SetValue<T>) {
    const previous = getValue()
    cache.current = isSetValueFunction(value) ? value(previous) : value
  }

  // const schedule = createScheduler<SetValue<T>>({
  //   onFinish() {
  //     cache.current = getValue()
  //     if (!canUpdate(cache, isEqual)) {
  //       return
  //     }
  //     state.emitter.emit()
  //   },
  //   onResolveItem: resolveValue,
  // })

  const state: RawState<T> = function () {
    const stateValue = getValue()
    const ctx = context.use()
    // console.log('CTX', ctx?.id, 'STATE', state.id)
    if (ctx && !state.emitter.contains(ctx.sub)) {
      ctx.addEmitter(state.emitter)
    }
    return stateValue
  }
  state.listen = function (listener: (value: T) => void) {
    return state.emitter.subscribe(() => {
      const final = cache.current
      if (isUndefined(final)) {
        throw new Error('The value is undefined')
      }
      listener(final)
    })
  }
  state.emitter = createEmitter<T>(() => state())
  state.id = generateId()

  const clearScheduler = createScheduler.add(state.id, {
    onFinish() {
      cache.current = getValue()
      if (!canUpdate(cache, isEqual)) {
        return
      }
      state.emitter.emit()
    },
    onResolveItem: resolveValue,
  })
  state.set = function (value) {
    createScheduler.schedule(state.id, value)
  }

  state.destroy = function () {
    cache.current = undefined
    getValue()
    clearScheduler()
    state.emitter.clear()
  }
  state.withName = function (name: string) {
    state.stateName = name
    return state
  }

  return state
}
