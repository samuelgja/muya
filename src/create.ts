import { createContext } from './create-context'
import type { Emitter } from './create-emitter'
import { createEmitter } from './create-emitter'
import { isAbortError, isFunction, isPromise } from './is'
import { cancelablePromise } from './common'
import type { IsEqual, Set, SetValue } from './types'
import { createBatcher } from './batch'

const context = createContext<StateNotGeneric | undefined>(undefined)

type DefaultValue<T> = T | (() => T)

interface StateNotGeneric {
  get: () => unknown
  reset: () => void
  value?: unknown
  id: string
  subscribe: (listener: (value: unknown) => void) => () => void
  remove: () => void
}
export interface GetState<T> extends StateNotGeneric {
  <S>(selector?: (stateValue: T) => S, isEqual?: IsEqual<S>): undefined extends S ? T : S
  emitter: Emitter<T>
  get: () => T
  reset: () => void
  value?: T
  id: string
  subscribe: (listener: (value: T) => void) => () => void
}
interface State<T> extends GetState<T> {
  set: Set<T>
}

// it still can be promise, this just handle if defaultValue is function
export function getDefaultValue<T>(state: State<T>, defaultValue: DefaultValue<T>): T {
  const value = isFunction(defaultValue) ? context.run(state, defaultValue) : defaultValue
  return value as T
}
let id = 0
function getId() {
  id++
  return id.toString(36)
}

export function create<T extends () => T>(defaultValue: DefaultValue<T>): GetState<ReturnType<T>>
export function create<T>(defaultValue: DefaultValue<T>): State<T>
export function create<T>(defaultValue: DefaultValue<T>): GetState<T> | State<T> {
  const contextSubscriptions = new Map<string, () => void>()
  let previousData: unknown
  const state: State<T> = <S>(
    selector: (stateValue: T) => S = (stateValue) => stateValue as unknown as S,
    isEqual: IsEqual<S> = (a, b) => a === b,
  ): undefined extends S ? T : S => {
    const ctx = context.use()
    const hasCtx = !!ctx
    if (!hasCtx) {
      throw new Error(`calling states must be only inside of a running context.`)
    }

    if (!contextSubscriptions.has(ctx.id)) {
      const unsubscribe = state.emitter.subscribe(() => {
        console.log('Calling reset from ', state.id)
        console.log('For derive context: ', ctx.id)
        ctx.reset()
      })
      contextSubscriptions.set(ctx.id, unsubscribe)
    }
    const selectedValue = selector(getValue())

    if (previousData !== undefined && isEqual(previousData as S, selectedValue)) {
      return previousData as undefined extends S ? T : S
    }
    previousData = selectedValue as undefined

    return selectedValue as undefined extends S ? T : S
  }

  const batch = createBatcher<T>({
    getValue,
    setValue: setValueRaw,
    onFlush: (current) => {
      console.log('flush', state.id, current)
      state.value = current
      state.emitter.emit()
    },
  })

  state.emitter = createEmitter(getValue)

  let isResolving = false
  function resolveValue(): T {
    if (isResolving) {
      return state.value as T
    }
    const value = getDefaultValue(state, defaultValue)
    if (!isPromise<T>(value)) {
      return value
    }

    isResolving = true
    const { promise, controller } = cancelablePromise<T>(value, batch.abortController)
    batch.abortController = controller
    promise
      .then((resolvedValue) => {
        isResolving = false
        state.value = resolvedValue
        state.emitter.emit()
      })
      .catch((error) => {
        isResolving = false

        if (isAbortError(error)) {
          return
        }
      })
    // switch Promise to cancelable promise
    return promise as unknown as T
  }

  function getValue() {
    if (state.value === undefined) {
      state.value = resolveValue()
    }
    return state.value
  }

  state.reset = function call() {
    state.value = resolveValue()
    state.emitter.emit()
  }

  state.get = getValue
  state.id = getId()
  state.remove = function remove() {
    for (const [, unsubscribe] of contextSubscriptions) {
      unsubscribe()
    }
    contextSubscriptions.clear()
    state.emitter.clear()
  }

  if (isFunction(defaultValue)) {
    return state
  }
  state.subscribe = function subscribe(listener: (value: T) => void) {
    return state.emitter.subscribe(() => {
      listener(state.get())
    })
  }

  // function setState(value: SetValue<T>) {
  //   if (state.value === undefined) {
  //     state.value = resolveValue()
  //   }

  //   if (!isSetValueFunction(value)) {
  //     if (batch.abortController) {
  //       batch.abortController.abort()
  //     }

  //     state.value = value
  //     state.emitter.emit()
  //     return
  //   }

  //   const result = value(state.value as PromiseAndValue<T>)
  //   if (!isPromise(result)) {
  //     state.value = result as T
  //     state.emitter.emit()
  //     return
  //   }
  //   result
  //     .then((resolvedValue) => {
  //       // check if state.value is resolved value
  //       if (isPromise(state.value) && batch.abortController) {
  //         batch.abortController.abort()
  //       }
  //       state.value = resolvedValue as T
  //       state.emitter.emit()
  //     })
  //     .catch((error) => {
  //       if (isAbortError(error)) {
  //         return
  //       }
  //     })
  // }

  function setValueRaw(value: T) {
    state.value = value
    state.emitter.emit()
  }

  function setStateBatch(value: SetValue<T>) {
    // setState(value)
    // console.log(value)
    batch.addValue(value)
  }

  state.set = setStateBatch

  return state
}
