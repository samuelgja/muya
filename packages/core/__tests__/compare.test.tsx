import { act, renderHook } from '@testing-library/react-hooks'
import { create } from '../create'
import { longPromise } from './test-utils'
import { waitFor } from '@testing-library/react'
import { atom, useAtom, useSetAtom } from 'jotai'

describe('compare', () => {
  it('should render async value with sync selector', async () => {
    const state = create(longPromise(100))
    const selectedState = state.select((value) => {
      return value + 2
    })
    const listen = jest.fn()

    const { result } = renderHook(() => {
      listen()
      const value = selectedState()
      return value
    })

    await waitFor(() => {
      expect(result.current).toBe(2)
      expect(listen).toHaveBeenCalledTimes(2)
    })

    state.set(1)

    await waitFor(() => {
      expect(result.current).toBe(3)
      // muya re-render only 3 times
      expect(listen).toHaveBeenCalledTimes(3)
    })
  })
  it('should render async value with sync selector with jotai', async () => {
    const state = atom(longPromise(100))

    const selectedState = atom(async (get) => {
      const value = get(state)
      return (await value) + 2
    })

    const listen = jest.fn()
    const { result: setResult } = renderHook(() => {
      const value = useSetAtom(state)
      return value
    })
    const { result } = renderHook(() => {
      listen()
      const value = useAtom(selectedState)
      return value
    })

    await waitFor(() => {
      expect(result.current[0]).toBe(2)
      expect(listen).toHaveBeenCalledTimes(3)
    })

    act(() => {
      setResult.current(1 as never)
    })

    await waitFor(() => {
      expect(result.current[0]).toBe(3)
      // OH it render 5 times, sad.
      expect(listen).toHaveBeenCalledTimes(5)
    })
  })
})
