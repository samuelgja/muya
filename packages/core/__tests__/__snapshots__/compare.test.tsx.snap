// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should count re-renders should compare counter: Jotai's counter 1`] = `0`;

exports[`should count re-renders should compare counter: Jotai's re-renders 1`] = `
[
  [],
  [],
]
`;

exports[`should count re-renders should compare counter: Jotai's re-renders 2`] = `
[
  [],
  [],
]
`;

exports[`should count re-renders should compare counter: Muya's counter 1`] = `0`;

exports[`should count re-renders should compare counter: Muya's re-renders 1`] = `
[
  [],
]
`;

exports[`should count re-renders should compare counter: Muya's re-renders 2`] = `
[
  [],
]
`;
